File: native/patch_xcode.py
#!/usr/bin/env python3
"""Patches the Xcode project to include TorrentFlowNativeService.swift."""

import os
import re
import uuid

PBXPROJ_PATH = os.path.join(
    os.path.dirname(__file__), '..', 'ios', 'Runner.xcodeproj', 'project.pbxproj')

SWIFT_FILE = ('TorrentFlowNativeService.swift', 'sourcecode.swift')


def generate_uuid():
    return uuid.uuid4().hex.upper()[:24]


def patch_xcode_project():
    with open(PBXPROJ_PATH, 'r', encoding='utf-8') as f:
        content = f.read()

    fname, file_type = SWIFT_FILE

    if f'/* {fname} */' in content:
        print(f"Already in project: {fname}")
        return

    # Locate Runner group and Sources phase
    group = re.search(r'([A-F0-9]{24})\s+/\*\s*Runner\s*\*/', content)
    phase = re.search(
        r'/\*\s*Begin PBXSourcesBuildPhase section.*?([A-F0-9]{24})\s+/\*\s*Sources\s*\*/',
        content, re.DOTALL)

    if not group or not phase:
        print("ERROR: Could not find Runner group or Sources phase")
        return

    group_id = group.group(1)
    phase_id = phase.group(1)
    ref_id = generate_uuid()
    file_id = generate_uuid()

    # Add PBXFileReference
    content = re.sub(
        r'(/\* End PBXFileReference section \*/)',
        f"""\t\t{ref_id} /* {fname} */ = {{
\t\t\tisa = PBXFileReference;
\t\t\tlastKnownFileType = {file_type};
\t\t\tname = "{fname}";
\t\t\tpath = "{fname}";
\t\t\tsourceTree = "<group>";
\t\t}};
\\1""", content, count=1)

    # Add PBXBuildFile
    content = re.sub(
        r'(/\* End PBXBuildFile section \*/)',
        f"""\t\t{file_id} /* {fname} in Sources */ = {{
\t\t\tisa = PBXBuildFile;
\t\t\tfileRef = {ref_id} /* {fname} */;
\t\t}};
\\1""", content, count=1)

    # Add to Runner group children
    content = re.sub(
        rf'({group_id}\s+/\*\s*Runner\s*\*/.*?children\s*=\s*\()',
        lambda m: m.group(1) + f'\n\t\t\t\t{ref_id} /* {fname} */,',
        content, count=1, flags=re.DOTALL)

    # Add to Sources build phase files
    content = re.sub(
        rf'({phase_id}\s+/\*\s*Sources\s*\*/.*?files\s*=\s*\()',
        lambda m: m.group(1) + f'\n\t\t\t\t{file_id} /* {fname} in Sources */,',
        content, count=1, flags=re.DOTALL)

    with open(PBXPROJ_PATH, 'w', encoding='utf-8') as f:
        f.write(content)

    print(f"Added {fname} to Xcode project")


if __name__ == '__main__':
    print("Patching Xcode project...")
    patch_xcode_project()
    print("Done!")

File: native/patch_xcode.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Patches the Xcode project to include TorrentFlowNativeService.swift

begin
  require 'xcodeproj'
rescue LoadError
  puts "Installing xcodeproj gem..."
  system('gem install xcodeproj --no-document') || exit(1)
  require 'xcodeproj'
end

PROJECT_PATH = File.join(__dir__, '..', 'ios', 'Runner.xcodeproj')
SWIFT_FILE = 'TorrentFlowNativeService.swift'

project = Xcodeproj::Project.open(PROJECT_PATH)

# Check if already exists
if project.main_group.find_subpath(File.join('Runner', SWIFT_FILE), false)
  puts "Already in project: #{SWIFT_FILE}"
  exit 0
end

# Find Runner group and target
runner_group = project.main_group.find_subpath('Runner', false) || exit(1)
target = project.targets.find { |t| t.name == 'Runner' } || exit(1)

# Add file reference and add to Sources build phase
file_ref = runner_group.new_file(SWIFT_FILE)
target.source_build_phase.add_file_reference(file_ref)

project.save
puts "Added #{SWIFT_FILE} to Xcode project"
